rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,266 @@
+
+Support for new configuration options will be added to Teleport's configuration
+file in order to enable and configure this feature.
+
+```yaml
+teleport:
+
+  # This section should be used on nodes which will join the cluster with the
+  # new aws join method, in place of auth_token.
+  aws_token:
+    enabled: true
+
+auth_service:
+
+  # This section should be used on auth servers which will allow nodes to
+  # join the cluster with the new aws join method.
+  aws_join:
+    # Deny rules will be checked first. If any deny rule matches an incoming
+    # node, it will be rejected.
+    deny:
+    - organization: "" # if organization is empty or omitted it matches any org
+      account: "" # if account is empty or omitted it matches any account
+
+    # Allow rules will be checked after deny rules. Incoming nodes will be
+    # accepted if they match any allow rule.
+    allow:
+    - organization: "" # if organization is empty or omitted it matches any org
+      account: "" # if accounts is empty or omitted it matches any account
+
+    # Example:
+    allow:
+    - account: "2222222222" # allow any node from this account
+    - organization: "o-1111111111" # allow any node in any account this org
+    deny:
+    - account: "3333333333" # this specific account in org "o-1111111111" should be rejected
+
+    # In theory we could add support for more claims, like "role" or "AMI",
+    # to be combined with the initially supported "organization" and "identity" claims
+```
+
+### AWS Configuration
+
+In all AWS accounts where nodes using the new aws join method will be
+deployed, you must create an IAM role with the following attached
+policy. All EC2 instances which will run Teleport nodes using the `ec2`
+credential source must be launched with this IAM role.
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "sts:GetCallerIdentity",
+                "organizations:DescribeOrganization"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+
+Notably, the credentials need only be accessible from the Teleport nodes, not
+the auth server. All requests will be signed on the nodes and the auth server
+will only forward them to public AWS endpoints. The auth server does not even
+need to run on AWS.
+
+### Alternatives
+
+I originally looked into doing something like Vault's [EC2 auth method](https://www.vaultproject.io/docs/auth/aws#ec2-auth-method).
+
+At a high level, instead of a join token or signed API requests, the node would
+present it's signed AWS Instance Identity Document. We could check the
+signature on this and the instance details to confirm the node's AWS account.
+
+The drawback of this method is that there is no information about the AWS
+Organization in the Identity Document, and being able to accept nodes from a
+given Organization rather than explicitly listing every account is a
+requirement for this feature.
```

Comment:
Not 100% sure. I think the main requirement is to simplify the flow. Whitelisting account should be enough if we don't have any security concerns. Need to double check this requirement.